csrc/scheduler/matmul_utils.cpp

```diff
@@ -187,10 +195,148 @@ std::string isMatmulFusionDefinitionSupported(
     // Check if MmaOp represents gemm (requires M/N/K == 1, B == 0)
     //  or bgemm (requires M/N/K/B == 1)
     constexpr size_t expected_axes_numbers = 1;
-    if (mma_expr->mAxes().size() != expected_axes_numbers ||
-        mma_expr->nAxes().size() != expected_axes_numbers ||
-        mma_expr->kAxes().size() != expected_axes_numbers ||
-        mma_expr->batchAxes().size() > expected_axes_numbers) {
+    if (props.empty()) {
+      if (mma_expr->mAxes().size() != expected_axes_numbers ||
+          mma_expr->nAxes().size() != expected_axes_numbers ||
+          mma_expr->kAxes().size() != expected_axes_numbers ||
+          mma_expr->batchAxes().size() > expected_axes_numbers) {
+        return "MmaOp has unsupported number of one of M/N/K/Batch axes";
+      }
+    } else {
+      MmaOpUtils::MmaOpDetails mma_details = MmaOpUtils::getMmaOpDetails(
+          props.front().out, props.front().a, props.front().b);
+      if (mma_details.m_axes.size() != expected_axes_numbers ||
+          mma_details.n_axes.size() != expected_axes_numbers ||
+          mma_details.k_axes.size() != expected_axes_numbers ||
+          mma_details.batch_axes.size() > expected_axes_numbers) {
+        return "MmaOp which can replace mul-sum pair has unsupported number of one of M/N/K/Batch axes";
+      }
+    }
+
+    if (!mma_output->hasReduction()) {
+      return "MMA output TV has no reduction domain";
+    }
+  }
+
+  // Quick checks - Fusion
+  {
+    // Fusion should contain at least two inputs (for now)
+    if (minimal_number_of_inputs > fusion_inputs.size()) {
+      return "Fusion inputs contain at least one non-TensorView object";
+    }
```

Comment:
Do we not support squaring square matrices?

Reply:
This was existing logic (and code) I moved around.